src/Ormolu/Fixity.hs

```diff
@@ -0,0 +1,247 @@
+
+-- | Build a fixity map using the given popularity threshold and a list of
+-- cabal dependencies. Dependencies from the list have higher priority than
+-- other packages.
+buildFixityMap ::
+  -- | Explicitely known dependencies
+  [String] ->
+  -- | Popularity ratio threshold, after which a very popular package will
+  -- completely rule out conflicting definitions coming from other packages
+  -- instead of being merged with them
+  Float ->
+  -- | Resulting map
+  FixityMap
+buildFixityMap = buildFixityMap' packageToOps packageToPopularity bootPackages
+
+-- | Build a fixity map using the given popularity threshold and a list of
+-- cabal dependencies. Dependencies from the list have higher priority than
+-- other packages. This specific version of the function allows the user to
+-- specify the package databases used to build the final fixity map.
+buildFixityMap' ::
+  -- | Map from package to fixity map for operators defined in this package
+  Map String FixityMap ->
+  -- | Map from package to popularity
+  Map String Int ->
+  -- | Higher priority packages
+  [String] ->
+  -- | Explicitely known dependencies
+  [String] ->
+  -- | Popularity ratio threshold, after which a very popular package will
+  -- completely rule out conflicting definitions coming from other packages
+  -- instead of being merged with them
+  Float ->
+  -- | Resulting map
+  FixityMap
+buildFixityMap'
+  operatorMap
+  popularityMap
+  higherPriorityPackages
+  dependencies
+  strategyThreshold =
+    -- Map.union is left biaised
+    Map.union baseFixityMap $
+      Map.union cabalFixityMap $
+        Map.union higherPriorityFixityMap remainingFixityMap
+    where
+      baseFixityMap =
+        Map.insert ":" colonFixityInfo $
+          fromMaybe Map.empty $
+            Map.lookup "base" operatorMap
+      cabalFixityMap =
+        mergeAll (buildPackageFixityMap <$> dependencies)
+      higherPriorityFixityMap =
+        mergeAll (buildPackageFixityMap <$> higherPriorityPackages)
+      remainingFixityMap =
+        mergeFixityMaps
+          popularityMap
+          strategyThreshold
+          (buildPackageFixityMap <$> remainingPackages)
+      remainingPackages =
+        Map.keys operatorMap
+          `difference` (dependencies ++ higherPriorityPackages)
+      buildPackageFixityMap packageName =
+        ( packageName,
+          fromMaybe Map.empty $
+            Map.lookup packageName operatorMap
+        )
+      difference l1 l2 =
+        Set.toList $
+          Set.fromList l1 `Set.difference` Set.fromList l2
+      mergeAll = mergeFixityMaps Map.empty 10.0
+
+-- | Merge a list of individual fixity maps, coming from different packages.
+-- Package popularities and the given threshold are used to choose between
+-- the "keep best only" (>= threshold) and "merge all" (< threshold)
+-- strategies when conflicting definitions are encountered for an operator.
+mergeFixityMaps ::
+  -- | Map from package to popularity
```

Comment:
Here and in the definition of `HackageInfo` it should be explained what `Int` means.